// @mui/utils/legacy/useLocalStorageState/useLocalStorageState.js
'use client';

import * as React from 'react';
var NOOP = function NOOP() {};

// storage events only work across tabs, we'll use an event emitter to announce within the current tab
var currentTabChangeListeners = new Map();
function onCurrentTabStorageChange(key, handler) {
  var listeners = currentTabChangeListeners.get(key);
  if (!listeners) {
    listeners = new Set();
    currentTabChangeListeners.set(key, listeners);
  }
  listeners.add(handler);
}
function offCurrentTabStorageChange(key, handler) {
  var listeners = currentTabChangeListeners.get(key);
  if (!listeners) {
    return;
  }
  listeners.delete(handler);
  if (listeners.size === 0) {
    currentTabChangeListeners.delete(key);
  }
}
function emitCurrentTabStorageChange(key) {
  var listeners = currentTabChangeListeners.get(key);
  if (listeners) {
    listeners.forEach(function (listener) {
      return listener();
    });
  }
}
function subscribe(area, key, cb) {
  var storageHandler = function storageHandler(event) {
    if (event.storageArea === area && event.key === key) {
      cb();
    }
  };
  window.addEventListener('storage', storageHandler);
  onCurrentTabStorageChange(key, cb);
  return function () {
    window.removeEventListener('storage', storageHandler);
    offCurrentTabStorageChange(key, cb);
  };
}
function getSnapshot(area, key) {
  return area.getItem(key);
}
function setValue(area, key, value) {
  if (typeof window !== 'undefined') {
    if (value === null) {
      area.removeItem(key);
    } else {
      area.setItem(key, String(value));
    }
    emitCurrentTabStorageChange(key);
  }
}
function useLocalStorageStateServer(key) {
  var initializer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  var _React$useState = React.useState(initializer),
    initialValue = _React$useState[0];
  return [initialValue, function () {}];
}

/**
 * Sync state to local storage so that it persists through a page refresh. Usage is
 * similar to useState except we pass in a storage key so that we can default
 * to that value on page load instead of the specified initial value.
 *
 * Since the storage API isn't available in server-rendering environments, we
 * return initialValue during SSR and hydration.
 *
 * Things this hook does different from existing solutions:
 * - SSR-capable: it shows initial value during SSR and hydration, but immediately
 *   initializes when clientside mounted.
 * - Sync state across tabs: When another tab changes the value in the storage area, the
 *   current tab follows suit.
 */

function useLocalStorageStateBrowser(key) {
  var initializer = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  var _React$useState2 = React.useState(initializer),
    initialValue = _React$useState2[0];
  var area = window.localStorage;
  var subscribeKey = React.useCallback(function (cb) {
    return key ? subscribe(area, key, cb) : NOOP;
  }, [area, key]);
  var getKeySnapshot = React.useCallback(function () {
    var _ref;
    return (_ref = key && getSnapshot(area, key)) != null ? _ref : initialValue;
  }, [area, initialValue, key]);
  var getKeyServerSnapshot = React.useCallback(function () {
    return initialValue;
  }, [initialValue]);
  var storedValue = React.useSyncExternalStore(subscribeKey, getKeySnapshot, getKeyServerSnapshot);
  var setStoredValue = React.useCallback(function (value) {
    if (key) {
      var valueToStore = value instanceof Function ? value(storedValue) : value;
      setValue(area, key, valueToStore);
    }
  }, [area, key, storedValue]);
  var _React$useState3 = React.useState(initialValue),
    nonStoredValue = _React$useState3[0],
    setNonStoredValue = _React$useState3[1];
  if (!key) {
    return [nonStoredValue, setNonStoredValue];
  }
  return [storedValue, setStoredValue];
}
export default typeof window === 'undefined' ? useLocalStorageStateServer : useLocalStorageStateBrowser;
